File: src/main/kotlin/com/linqingying/cangjie/psi/CjAnnotationEntry.java
package com.linqingying.cangjie.psi;

import com.linqingying.cangjie.CjNodeTypes;
import com.linqingying.cangjie.name.Name;
import com.linqingying.cangjie.psi.stubs.CangJieAnnotationEntryStub;
import com.linqingying.cangjie.psi.stubs.elements.CjStubElementTypes;
import com.intellij.lang.ASTNode;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Collections;
import java.util.List;

public class CjAnnotationEntry extends CjElementImplStub<CangJieAnnotationEntryStub> implements CjCallElement  {


    public CjAnnotationEntry(@NotNull ASTNode node) {
        super(node);
    }

    public CjAnnotationEntry(@NotNull CangJieAnnotationEntryStub stub) {
        super(stub, CjStubElementTypes.ANNOTATION_ENTRY);
    }
    @Override
    public CjConstructorCalleeExpression getCalleeExpression() {
        return getStubOrPsiChild(CjStubElementTypes.CONSTRUCTOR_CALLEE);
    }

    @Override
    public @NotNull List<CjLambdaArgument> getLambdaArguments() {
        return Collections.emptyList();

    }

    @Override
    public @NotNull List<CjTypeProjection> getTypeArguments() {
        CjTypeArgumentList typeArgumentList = getTypeArgumentList();
        if (typeArgumentList == null) {
            return Collections.emptyList();
        }
        return typeArgumentList.getArguments();
    }

    @Override
    public @Nullable CjTypeArgumentList getTypeArgumentList() {
        return null;
    }

    @Override
    public CjValueArgumentList getValueArgumentList() {
        CangJieAnnotationEntryStub stub = getStub();
        if (stub == null && getGreenStub() != null) {
            return findChildByType(CjNodeTypes.VALUE_ARGUMENT_LIST);
        }

        return getStubOrPsiChild(CjStubElementTypes.VALUE_ARGUMENT_LIST);
    }

    @Override
    public @NotNull List<? extends ValueArgument> getValueArguments() {
        CangJieAnnotationEntryStub stub = getStub();
        if (stub != null && !stub.hasValueArguments()) {
            return Collections.<CjValueArgument>emptyList();
        }

        CjValueArgumentList list = getValueArgumentList();
        return list != null ? list.getArguments() : Collections.<CjValueArgument>emptyList();
    }

    @Nullable @IfNotParsed
    public CjTypeReference getTypeReference() {
        CjConstructorCalleeExpression calleeExpression = getCalleeExpression();
        if (calleeExpression == null) {
            return null;
        }
        return calleeExpression.getTypeReference();
    }
    @Nullable
    public Name getShortName() {
      CangJieAnnotationEntryStub stub = getStub();
        if (stub != null) {
            String shortName = stub.getShortName();
            if (shortName != null) {
                return Name.identifier(shortName);
            }
            return null;
        }

        CjTypeReference typeReference = getTypeReference();
        assert typeReference != null : "Annotation entry hasn't typeReference " + getText();
       CjTypeElement typeElement = typeReference.getTypeElement();
        if (typeElement instanceof CjUserType userType) {
            String shortName = userType.getReferencedName();
            if (shortName != null) {
                return Name.identifier(shortName);
            }
        }
        return null;
    }


}

File: src/main/kotlin/com/linqingying/cangjie/psi/CjAnnotationEntry.kt
/*
 * Copyright 2024 LinQingYing. and contributors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * The use of this source code is governed by the Apache License 2.0,
 * which allows users to freely use, modify, and distribute the code,
 * provided they adhere to the terms of the license.
 *
 * The software is provided "as-is", and the authors are not responsible for
 * any damages or issues arising from its use.
 *
 */

package com.linqingying.cangjie.psi

import com.linqingying.cangjie.CjNodeTypes
import com.linqingying.cangjie.lexer.CjTokens
import com.linqingying.cangjie.name.Name
import com.linqingying.cangjie.name.Name.Companion.identifier
import com.linqingying.cangjie.psi.stubs.CangJieAnnotationEntryStub
import com.linqingying.cangjie.psi.stubs.elements.CjStubElementTypes
import com.intellij.lang.ASTNode
import com.intellij.psi.PsiElement

class CjAnnotationEntry : CjElementImplStub<CangJieAnnotationEntryStub >, CjCallElement {
    constructor(node: ASTNode) : super(node)

    constructor(stub: CangJieAnnotationEntryStub) : super(stub, CjStubElementTypes.ANNOTATION_ENTRY)

    override val calleeExpression : CjConstructorCalleeExpression? get(){
        return getStubOrPsiChild(CjStubElementTypes.CONSTRUCTOR_CALLEE)
    }

    override val lambdaArguments: List<CjLambdaArgument> get(){
        return emptyList()
    }

    override val typeArguments : List<CjTypeProjection> get(){
        val typeArgumentList = typeArgumentList ?: return emptyList()
        return typeArgumentList.arguments
    }

    override val typeArgumentList : CjTypeArgumentList? = null

    override val valueArgumentList : CjValueArgumentList? get(){
        val stub = stub
        if (stub == null && greenStub != null) {
            return findChildByType(CjNodeTypes.VALUE_ARGUMENT_LIST)
        }

        return getStubOrPsiChild(CjStubElementTypes.VALUE_ARGUMENT_LIST)
    }

    override val valueArguments : List<ValueArgument > get(){
        val stub = stub
        if (stub != null && !stub.hasValueArguments()) {
            return emptyList<CjValueArgument>()
        }

        val list = valueArgumentList
        return list?.arguments ?: emptyList<CjValueArgument>()
    }
    val atSymbol : PsiElement? get()    {
        return findChildByType(CjTokens.AT)
    }
    @get:IfNotParsed
    val typeReference: CjTypeReference?
        get() {
            val calleeExpression = calleeExpression ?: return null
            return calleeExpression.typeReference
        }
    val shortName: Name?
        get() {
            val stub = stub
            if (stub != null) {
                val shortName = stub.getShortName()
                if (shortName != null) {
                    return identifier(shortName)
                }
                return null
            }
            typeReference ?: return null
            val typeReference =
                checkNotNull(typeReference) { "Annotation entry hasn't typeReference $text" }
            val typeElement = typeReference.typeElement
            if (typeElement is CjUserType) {
                val shortName = typeElement.referencedName
                if (shortName != null) {
                    return identifier(shortName)
                }
            }
            return null
        }

}
